### PR TITLE
#1777 - Process MSFAA Cancellation and Reactivation Notification

### PR DESCRIPTION
**Pull request by sh16011993**
**As a part of this PR:**

- Created a GC Notify template to send emails when a MSFAA record is cancelled.
- Updated msfaa-number service to save notifications on MSFAA cancellation.
- MSFAA reactivation template is not required as the federal government will send them notification. 

Screenshot of the sent notification email:

image

**To be done in a separate PR:**

- E2E tests for MSFAA Cancellation Notification.

**NOTE:** There is an update in the Acceptance Criteria (point 2) - MSFAA reactivation template is not required as the federal government will send them notification. Only MSFAA cancellation notification needs to be sent.

### CODE REVIEW

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 1]
```diff
@@ -191,22 +196,56 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
...
+        relations: {
+          student: { user: true },
+        },
+        where: {
+          msfaaNumber,
+        },
+      });
+      const systemUser = await this.systemUsersService.systemUser();
+      // Only if the msfaa record was successfully cancelled, then save the msfaa cancellation notification.
+      // This also takes care of the scenario where the msfaa record was already cancelled before in which case the record will not be update and hence no notification will be saved.
```
Comment (guru-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:236]: Please change it to "update and hence no notification will be sent"

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 2]
```diff
@@ -191,22 +196,56 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
     cancelledDate: Date,
     newIssuingProvince: string,
   ): Promise<UpdateResult> {
-    if (!cancelledDate || !newIssuingProvince) {
-      throw new Error(
-        "Not all required fields to update a received MSFAA record were provided.",
+    return this.dataSource.transaction(async () => {
+      // Keeping both MSFAA Cancellation update and saving of MSFAA Notification message as a part of the same transaction.
+      if (!cancelledDate || !newIssuingProvince) {
```
Comment (ann-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:201]: do we need this check inside transaction?

File: sources/packages/backend/libs/sims-db/src/entities/notification.model.ts [hunk 3]
```diff
@@ -132,4 +132,8 @@ export enum NotificationMessageType {
    * ECE response file processing details.
    */
   ECEResponseFileProcessing = 11,
+  /**
+   * MSFAA record gets cancelled.
+   */
+  MSFAACancellationComplete = 12,
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/notification.model.ts:138]: What about `MSFAACancellation` for enum?

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 4]
```diff
@@ -191,22 +196,56 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
...
+        throw new Error(
+          "Not all required fields to update a received MSFAA record were provided.",
+        );
+      }
+      // Update the MSFAA record that needs to be cancelled.
+      const updateResult = await this.repo.update(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:207]: Please review how other transactions are executed, usually, they are declared as below.
`await this.dataSource.transaction(async (transactionalEntityManager)`
If you was able to check that the current code is also working and the codes are part of the same transaction, can you please provide further information?

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 5]
```diff
@@ -191,22 +196,56 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
...
+        },
+        where: {
+          msfaaNumber,
+        },
+      });
+      const systemUser = await this.systemUsersService.systemUser();
+      // Only if the msfaa record was successfully cancelled, then save the msfaa cancellation notification.
+      // This also takes care of the scenario where the msfaa record was already cancelled before in which case the record will not be update and hence no notification will be saved.
+      if (updateResult.affected) {
+        this.notificationActionsService.saveMSFAACancellationNotification(
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:238]: This method also expected the `entityManager` to be part of an existing transaction. Please check the other methods using something like `this.notificationActionsService.saveSomething` to see how they were implemented.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 6]
```diff
@@ -6,7 +6,11 @@ import {
   OfferingIntensity,
 } from "@sims/sims-db";
 import { getISODateOnlyString } from "@sims/utilities";
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:8]: Are we not suppose to send notification when we cancel an MSFAA during re-activation? 
image

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:201]: The `if` can be moved outside the transaction block.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 4 above)
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:207]: May not be part of the PR, but we are not throwing error here on `(!updateResult.affected)` similar to `updateReceivedRecord()`. It will be nice to add that.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:8]: I would say no @dheepak-aot. In my understanding, we are deactivating some MSFAA(s) in order to activate other. The student will receive an email notification about it.
The point about sending a notification during the cancellation is that it may require an action from the student and it will not be communicated in any other way.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 7]
```diff
@@ -191,22 +196,56 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
...
-        cancelledDate: IsNull(),
-        newIssuingProvince: IsNull(),
-      },
-      {
-        cancelledDate: getISODateOnlyString(cancelledDate),
-        newIssuingProvince,
-      },
-    );
+      // Get the student associated with the cancelled MSFAA record.
+      const msfaaRecord = await this.repo.findOne({
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:219]: Please move the `findOne` inside the `if (updateResult.affected)` since it is only needed if the notification will be sent.

**Review by andrewsignori-aot — COMMENTED** [review]
Nice work, please take a look at the comments.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 8]
```diff
@@ -111,6 +112,42 @@ export class NotificationActionsService {
...
+    entityManager?: EntityManager,
+  ): Promise<void> {
+    const templateId = await this.notificationMessageService.getTemplateId(
+      NotificationMessageType.MSFAACancellationComplete,
+    );
+
+    const notificationToSend = {
+      userId: notification.userId,
+      messageType: NotificationMessageType.MSFAACancellationComplete,
+      messagePayload: {
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:133]: Please have the message payload typed like this. 

```ts
const messagePayload: NotificationEmailMessage = {
      email_address: notification.toAddress,
      template_id: templateId,
      personalisation: {
        givenNames: notification.givenNames ?? "",
        lastName: notification.lastName,
      },
    };

    const notificationToSend = {
      userId: notification.userId,
      messageType: NotificationMessageType.MSFAACancellationComplete,
      messagePayload: messagePayload,
    };
```

**Review by ann-aot — COMMENTED** [review]
Good work @sh16011993 . Pls take a look at the comments

**Comment by dheepak-aot** [comment]
Good job on first notification effort. Please have a look at the comments.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 1 above)
Comment (sh16011993) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:236]: I believe `saved` is right here because here we are saving the notifications only. Sending will happen through the `processNotifications` job later.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 2 above)
Comment (sh16011993) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:201]: Moved it outside. Thanks 👍

File: sources/packages/backend/libs/sims-db/src/entities/notification.model.ts (hunk 3 above)
Comment (sh16011993) [reply sources/packages/backend/libs/sims-db/src/entities/notification.model.ts:138]: Changed it. Thanks 👍

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 4 above)
Comment (sh16011993) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:207]: You are right. I missed to provide the ready-to-use `transactionalEntityManager` as an argument to this `transaction` callback. Corrected it. Thanks 👍

Comment (sh16011993) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:207]: Added it. Thanks 👍

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 9]
```diff
@@ -196,17 +201,58 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
...
+      );
+      // In case no MSFAA record was cancelled.
+      if (!updateResult.affected) {
+        throw new Error(
+          `Error while cancelling MSFAA number: ${msfaaNumber}. Number of affected rows was ${updateResult.affected}, expected 1.`,
+        );
+      }
+      // Only if the msfaa record was successfully cancelled, then save the msfaa cancellation notification.
+      // This also takes care of the scenario where the msfaa record was already cancelled before in which case the record will not be updated and hence no notification will be saved.
+      if (updateResult.affected) {
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:226]: This condition is not required anymore. If code is past line 224 the updateResult.affected cannot be 0.

Comment (sh16011993) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts]: Removed it. Thanks.

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts [hunk 10]
```diff
@@ -196,17 +201,58 @@ export class MSFAANumberService extends RecordDataModelService<MSFAANumber> {
...
+          cancelledDate: IsNull(),
+          newIssuingProvince: IsNull(),
+        },
+        {
+          cancelledDate: getISODateOnlyString(cancelledDate),
+          newIssuingProvince,
+        },
+      );
+      // In case no MSFAA record was cancelled.
+      if (!updateResult.affected) {
```
Comment (ann-aot) [inline sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:219]: we are throwing the error on the consumers function, so, do we need it here? 
image

Comment (sh16011993) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts]: I have removed it from the consumer to keep it with `msfaa-number.service.ts` generating the `updateResult`.

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes. Good work @sh16011993 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good!

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes. 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @sh16011993

File: sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/msfaa-number/msfaa-number.service.ts:207]: This must use the transaction manager. Please see other similar codes along the source code for reference.
image

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes. There is at least one pending comment. Please take a look.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, looks good 👍
